File: rust-api/src/cache.rs
//! In-memory TTL caching (moka).
//!
//! Provider calls scrape slow upstream pages, so identical requests must never
//! hit the Node workers twice within a short window. Every `/api/*` handler
//! keys its cache as `provider|endpoint|params` and serves the stored JSON on
//! a hit, stamping `Cache-Control: public, max-age=<ttl>` on the response.
//!
//! Capacities: each cache holds up to 10 000 entries with a per-endpoint TTL
//! configured by [`crate::config::Config`] (`CACHE_*_SECS`).

use std::sync::Arc;
use std::time::Duration;

use moka::future::Cache;
use serde_json::Value;

/// A single endpoint's TTL cache.
#[derive(Clone)]
pub struct TtlCache {
    inner: Cache<String, Arc<Value>>,
    ttl: Duration,
}

impl TtlCache {
    /// Build a cache with the given TTL (seconds, minimum 1).
    pub fn new(ttl_secs: u64) -> Self {
        Self {
            inner: Cache::builder()
                .time_to_live(Duration::from_secs(ttl_secs.max(1)))
                .max_capacity(10_000)
                .build(),
            ttl: Duration::from_secs(ttl_secs),
        }
    }

    /// Look up a key, returning a clone of the cached JSON if present.
    pub async fn get(&self, key: &str) -> Option<Value> {
        self.inner.get(key).await.map(|v| (*v).clone())
    }

    /// Store a value under a key.
    pub async fn set(&self, key: String, value: Value) {
        self.inner.insert(key, Arc::new(value)).await;
    }

    /// The configured TTL, used for the `Cache-Control` header.
    pub fn ttl(&self) -> Duration {
        self.ttl
    }
}

/// The bundle of per-endpoint caches held in [`crate::state::AppState`].
#[derive(Clone)]
pub struct CacheBundle {
    pub catalog: TtlCache,
    pub search: TtlCache,
    pub meta: TtlCache,
    pub episodes: TtlCache,
    pub stream: TtlCache,
}

impl CacheBundle {
    /// Construct all five caches from the configured TTLs.
    pub fn from_config(c: &crate::config::Config) -> Self {
        Self {
            catalog: TtlCache::new(c.cache_catalog_secs),
            search: TtlCache::new(c.cache_search_secs),
            meta: TtlCache::new(c.cache_meta_secs),
            episodes: TtlCache::new(c.cache_episodes_secs),
            stream: TtlCache::new(c.cache_stream_secs),
        }
    }

    /// Build the canonical cache key for an endpoint/provider/params triple.
    pub fn key(endpoint: &str, provider: &str, params: &serde_json::Value) -> String {
        format!("{endpoint}|{provider}|{params}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[actix_rt::test]
    async fn get_set_roundtrip() {
        let c = TtlCache::new(60);
        assert!(c.get("k").await.is_none());
        c.set("k".into(), json!({"x": 1})).await;
        assert_eq!(c.get("k").await, Some(json!({"x": 1})));
    }

    #[actix_rt::test]
    async fn set_overwrites() {
        let c = TtlCache::new(60);
        c.set("k".into(), json!(1)).await;
        c.set("k".into(), json!(2)).await;
        assert_eq!(c.get("k").await, Some(json!(2)));
    }

    #[test]
    fn key_format_is_stable() {
        let params = json!({"query": "inception", "page": 1});
        let k1 = CacheBundle::key("search", "vega", &params);
        let k2 = CacheBundle::key("search", "vega", &params);
        let k3 = CacheBundle::key("search", "showbox", &params);
        assert_eq!(k1, k2);
        assert_ne!(k1, k3);
        assert!(k1.starts_with("search|vega|"));
    }

    #[test]
    fn ttl_is_exposed() {
        assert_eq!(TtlCache::new(30).ttl(), Duration::from_secs(30));
    }
}

File: gateway/lib/cache.ts
// In-memory TTL caching with single-flight request coalescing.
//
// Provider calls scrape slow upstream pages, so identical requests must never
// hit the providers twice within a short window. Each entry is stored with its
// `expiresAt` and pruned lazily on read/write. When a miss happens while the
// same key is already being loaded, the waiting callers share the in-flight
// promise instead of re-fanning out (this is what makes concurrent identical
// requests cheap — the core of "handling many requests" well).

export interface CacheEntry {
  value: unknown;
  expiresAt: number;
}

/**
 * A single endpoint's TTL cache. The backing Map is intentionally module-
 * scoped per instance; nothing in the request path mutates global state.
 */
export class TtlCache {
  private readonly store = new Map<string, CacheEntry>();
  private readonly inflight = new Map<string, Promise<unknown>>();
  readonly ttlSecs: number;

  constructor(ttlSecs: number) {
    this.ttlSecs = ttlSecs;
  }

  private now(): number {
    return Date.now();
  }

  private purgeExpired(): void {
    const now = this.now();
    for (const [key, entry] of this.store) {
      if (entry.expiresAt <= now) this.store.delete(key);
    }
  }

  /** Look up a key, returning a clone of the cached value if present. */
  get(key: string): unknown | undefined {
    const entry = this.store.get(key);
    if (!entry) return undefined;
    if (entry.expiresAt <= this.now()) {
      this.store.delete(key);
      return undefined;
    }
    return entry.value;
  }

  /** Store a value under a key. */
  set(key: string, value: unknown, ttlSecs?: number): void {
    const ttl = Math.max(1, ttlSecs ?? this.ttlSecs);
    this.store.set(key, { value, expiresAt: this.now() + ttl * 1000 });
  }

  /**
   * Return the cached value or compute it via `loader`, caching the result.
   * Concurrent calls for the same key await the same in-flight computation.
   */
  async getOrSet(key: string, loader: () => Promise<unknown>): Promise<unknown> {
    const cached = this.get(key);
    if (cached !== undefined) return cached;

    const pending = this.inflight.get(key);
    if (pending) {
      // A sibling request is already computing this key — await it.
      return pending;
    }

    const promise = loader()
      .then((value) => {
        this.set(key, value);
        return value;
      })
      .catch((err: unknown) => {
        // Don't cache failures; let callers observe the error.
        throw err;
      })
      .finally(() => {
        this.inflight.delete(key);
      });

    this.inflight.set(key, promise);
    return promise;
  }

  /** Drop everything (used by tests and config reloads). */
  clear(): void {
    this.store.clear();
    this.inflight.clear();
  }

  get size(): number {
    this.purgeExpired();
    return this.store.size;
  }
}

/** The bundle of per-endpoint caches shared by all handlers. */
export class CacheBundle {
  readonly catalog: TtlCache;
  readonly search: TtlCache;
  readonly meta: TtlCache;
  readonly episodes: TtlCache;
  readonly stream: TtlCache;
  readonly searchAll: TtlCache;

  constructor() {
    this.catalog = new TtlCache(300);
    this.search = new TtlCache(60);
    this.meta = new TtlCache(60);
    this.episodes = new TtlCache(300);
    this.stream = new TtlCache(30);
    this.searchAll = new TtlCache(60);
  }

  /** Build the canonical cache key for an endpoint/provider/params triple. */
  static key(endpoint: string, provider: string, params: unknown): string {
    return `${endpoint}|${provider}|${JSON.stringify(params)}`;
  }
}
